Backport ARM64EC variadic args fixes to LLVM 18

dpaoliello opened the pull request:
Backports fixes for ARM64EC variadic args:
* <https://github.com/llvm/llvm-project/pull/80595>
* <https://github.com/llvm/llvm-project/pull/79774>
* <https://github.com/llvm/llvm-project/pull/80994>

bylaws committed:
[AArch64] Fix variadic tail-calls on ARM64EC (#79774)

ARM64EC varargs calls expect that x4 = sp at entry, special handling is
needed to ensure this with tail calls since they occur after the
epilogue and the x4 write happens before.

I tried going through AArch64MachineFrameLowering for this, hoping to
avoid creating the dummy object but this was the best I could do since
the stack info that uses isn't populated at this stage,
CreateFixedObject also explicitly forbids 0 sized objects.

bylaws committed:
[AArch64] Fix generated types for ARM64EC variadic entry thunk targets (#80595)

ISel handles filling in x4/x5 when calling variadic functions as they
don't correspond to the 5th/6th X64 arguments but rather to the end of
the shadow space on the stack and the size in bytes of all stack
parameters (ignored and written as 0 for calls from entry thunks).

Will PR a follow up with ISel handling after this is merged.

efriedma-quic commented:
I was sort of waiting until the discussion on #80994 resolves... we might end up reverting parts of #80595 .

I guess it won't do any harm to land as-is, though.

dpaoliello commented:
> I was sort of waiting until the discussion on #80994 resolves... we might end up reverting parts of #80595 .
> 
> I guess it won't do any harm to land as-is, though.

I'll wait - I'd rather not drop something into the release branch if we know it may cause issues.

bylaws committed:
[AArch64] Skip over shadow space for ARM64EC entry thunk variadic calls (#80994)

When in an entry thunk the x64 SP is passed in x4 but this cannot be
directly passed through since x64 varargs calls have a 32 byte shadow
store at SP followed by the in-stack parameters. ARM64EC varargs calls
on the other hand expect x4 to point to the first in-stack parameter.

dpaoliello commented:
@efriedma-quic I think this is ready to land, the MacOS test failure is a known issue.

efriedma-quic reviewed (commented):
Looks fine, but I think merging might need to wait for a point release?  CC @tstellar .

dpaoliello commented:
@efriedma-quic 18.1.0 was just released, so I think now's a good time to merge this in for 18.1.1

efriedma-quic reviewed (approved):
LGTM

dpaoliello commented:
@tstellar can we please get this merged into the v18 release branch?

tstellar commented:
Hi @dpaoliello  (or anyone else). If you would like to add a note about this fix in the release notes (completely optional). Please reply to this comment with a one or two sentence description of the fix.  When you are done, please add the release:note label to this PR.
